File: examples/cpp/binary_search.cpp
#include <iostream>
#include <vector>

int binarySearch(const std::vector<int>& numbers, int target) {
    int left = 0;
    int right = static_cast<int>(numbers.size()) - 1;

    while (left <= right) {
        int middle = left + (right - left) / 2;
        if (numbers[middle] == target) {
            return middle;
        }
        if (numbers[middle] < target) {
            left = middle + 1;
        } else {
            right = middle - 1;
        }
    }
    return -1;
}

int main() {
    const std::vector<int> numbers{1, 4, 7, 12, 18, 25};
    if (binarySearch(numbers, 12) != 3 ||
        binarySearch(numbers, 8) != -1) {
        return 1;
    }
    std::cout << "C++ binary search: OK\n";
}

File: examples/python/binary_search.py
def binary_search(numbers: list[int], target: int) -> int:
    left = 0
    right = len(numbers) - 1

    while left <= right:
        middle = left + (right - left) // 2
        if numbers[middle] == target:
            return middle
        if numbers[middle] < target:
            left = middle + 1
        else:
            right = middle - 1
    return -1


if __name__ == "__main__":
    values = [1, 4, 7, 12, 18, 25]
    assert binary_search(values, 12) == 3
    assert binary_search(values, 8) == -1
    print("Python binary search: OK")
